_posts/en/posts/2024-09-18-disclose-headers-oom.md

```diff
@@ -0,0 +1,112 @@
+---
+
+Before Bitcoin Core v24.0.1, attackers could spam nodes with low-difficulty headers chains, which
+could be used to remotely crash peers.
+
+This issue is considered **High** severity.
+
+## Details
+
+Bitcoin Core stores the blockchain headers in memory. This makes it susceptible to being DoSed, by
+having it download and store extremely long chains of headers, even if they are of low difficulty.
+It is important to note that once crafted, an attack chain could be reused to crash any node on the
+network.
+
+The possibility of using this to attack nodes has long been known, and was the primary reason why
+the checkpoint system was still in place: making an attacker start an attack at the last checkpoint
+makes it far more costly than starting at the genesis block. However, over time, with decreasing
+hashrate costs, even this mitigation became less effective.
+
+This attack was independently discovered and reported to the Bitcoin Core project in January 2019 by
+David Jaenson, who suggested introducing newer checkpoints as a practical mitigation. However:
+1. This still leaves nodes performing IBD with no protection before they receive checkpoint blocks.
+2. It relies on the ecosystem semi-regularly adopting updated software with new checkpoints, a
+   practice which Bitcoin Core contributors have long been uncomfortable with.
+
+It later got increased attention when Braydon Fuller [posted his "Chain width expansion"
+writeup](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2019-October/017354.html) to the
+bitcoin-dev mailing list in October 2019. He had previously responsibly reported it to the Bitcoin
+Core security list. The suggested approach was not adopted in Bitcoin Core due to concerns about
+network convergence when limiting the number of parallel chains.
+
+At the time, the computational cost of creating a huge low-difficulty headers chain was equal to
+about 32.28% of mining one block at the tip. That is a cost of about 4.12 BTC since the block reward
+then was about 12.77 BTC.
+
+By February 2022, the cost of the attack had dropped further to around 14.73% of the cost of mining
+a block, and this prompted investigation of alternative solutions. If unaddressed, the cost today
+(September 2024) would just be 4.44% of a block. These figures translate to a cost of about 1.07 BTC
+and 0.14 BTC respectively, given the block reward at these dates.
+
+A protection against this DoS was implemented in Bitcoin Core PR
+[#25717](https://github.com/bitcoin/bitcoin/pull/25717), whereby the node will first verify a
+presented chain has enough work before committing to store it. With that, Bitcoin Core no longer
+relies on having checkpoints to protect against any known attacks.
+
+## Attribution
+
+Credit goes to David Jaenson for independently re-discovering the attack, estimating its cost and
+suggesting modifications.
```

Comment:
From the timeline and above description, I’m wondering whether Braydon Fuller was also supposed to be mentioned here.

Reply:
Ugh, good catch, i dropped the "and Braydon Fuller" part somehow. Fixed.